### PR TITLE
[LETS-32] Add meta log class to store information about the log

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/LETS-32

Add a class to store persistent information about the log that cannot be saved in the log, e.g. the checkpoint information.

**cublog::meta**

  - Stores information from multiple checkpoints, offers interface to add, get, remove checkpoint information.
  - Is saved to disk & is loaded from disk through packer/unpacker utilities
  - Add an instance to log_Gl

**cublog::checkpoint_info**

  - Add default constructor
  - Inherit packable_object publicly to offer others access to the pack/unpack functions

**Unit test for cublog::meta**

  - Added test_meta_log executable
  - Test adding/removing checkpoint info
  - Test unloading/loading from file
  - For the purpose of this test, only checkpoint_info.m_start_redo_lsa field is used

**Unit test: others**

  - Added fake_packable_object.hpp/EXPAND_PACKABLE_OBJECT_EMPTY_DEF to add empty definitions of packable objects derived classes easily. Used to add meta/checkpoint_info dependencies.
  - Updated test_main_prior_list_serialize new dependencies. Also replaced error manager / system parameter definitions with the corresponding fake files.
  - Added ${TEST_FAKE_DIR} and ${TEST_COMMON_DIR}
  - Removed superfluous if(AT_LEAST_ONE_UNIT_TEST) check from unit_tests/CMakeLists.txt. The project is only added if(AT_LEAST_ONE_UNIT_TEST) is true.

### CODE REVIEW

**Review by adrianbrab — APPROVED** [review]

**Review by hornetmj — APPROVED** [review]

File: src/transaction/log_meta.cpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  {
+    auto find_it = m_checkpoints.find (checkpoint_lsa);
+    if (find_it == m_checkpoints.cend ())
+      {
+	return nullptr;
+      }
+    else
+      {
+	return &find_it->second;
+      }
```
Comment (danilailie) [inline src/transaction/log_meta.cpp:138]: `if (find_it != m_checkpoints.cend ())
      {
	return &find_it->second;
      }
return nullptr;
    `

**Review by danilailie — APPROVED** [review]

File: src/transaction/log_meta.hpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+      void load_from_file (std::FILE *stream);       // load meta from meta log file
+      void flush_to_file (std::FILE *stream) const;  // write meta to disk
+
+      size_t get_packed_size (cubpacking::packer &serializator, std::size_t start_offset = 0) const;
+      void pack (cubpacking::packer &serializator) const;
+      void unpack (cubpacking::unpacker &deserializator);
+
+      const checkpoint_info *get_checkpoint_info (const log_lsa &checkpoint_lsa) const;
+      void add_checkpoint_info (const log_lsa &chkt_lsa, checkpoint_info &&chkpt_info);
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/transaction/log_meta.hpp]: Correct :)

File: src/transaction/log_meta.cpp [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+#include <memory>
+
+namespace cublog
+{
+  size_t
+  meta::get_packed_size (cubpacking::packer &serializator, std::size_t start_offset) const
+  {
+    size_t size = 0;
+
+    size += serializator.get_packed_int_size (start_offset + size); // m_checkpoints.size ()
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/transaction/log_meta.cpp]: Outdated: replaced with end_offset.

File: src/transaction/log_meta.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    for (const auto chkinfo : m_checkpoints)
+      {
+	size += serializator.get_packed_bigint_size (start_offset + size);
+	size += serializator.get_packed_int_size (start_offset + size);
+	size += serializator.get_packed_size_overloaded (chkinfo.second, start_offset + size);
+      }
+    return size;
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/transaction/log_meta.cpp]: Good point :)

File: src/transaction/log_meta.cpp (hunk 1 above)
Comment (ribram) [reply src/transaction/log_meta.cpp]: Done.

**Pull request merged**